src/crypto/isa-l/md5/async_md5.h

```diff
@@ -0,0 +1,145 @@
+
+ private:
+  friend class Batch;
+  MD5_HASH_CTX ctx;
+  boost::asio::any_completion_handler<void(error_code)> handler;
+};
+
+/// An asynchronous hash manager that uses batching to take advantage of
+/// vectorized hash calculations. The best supported implementation of
+/// SSE/AVX/AVX2/AVX512 is selected at runtime.
+///
+/// Hash requests submitted via async_hash() may be delayed until a full batch
+/// is ready to process. The exact batch size is between 8 and 32, but depends
+/// on the implementation.
+///
+/// An optional batch_timeout can be specified in the constructor to force the
+/// processing of an incomplete batch instead of waiting indefinitely. Member
+/// function async_flush() can also be called to force the processessing of an
+/// incomplete batch.
+class Batch {
+ public:
+  using executor_type = boost::asio::any_io_executor;
+  executor_type get_executor() const noexcept { return ex; }
+
+  /// Construct the manager to run on the given executor.
+  Batch(const executor_type& ex,
+        std::chrono::nanoseconds batch_timeout);
+
+  /// Issue an asynchronous request to hash the input bytes and update the
+  /// digest. Pass last=true to indicate that the digest should be finalized.
+  /// After completion, the hex-encoded digest can be read via Digest::as_hex().
+  template <boost::asio::completion_token_for<void(error_code)> CompletionToken>
+  auto async_hash(Digest& digest, std::string_view input,
+                  bool last, CompletionToken&& token)
+  {
+    // maintain work on the default executor
+    auto consigned = boost::asio::consign(
+        std::forward<CompletionToken>(token),
+        boost::asio::make_work_guard(get_executor()));
+    return boost::asio::async_initiate<decltype(consigned), void(error_code)>(
+        initiate_async_hash{this}, consigned, std::ref(digest), input, last);
+  }
+
+  /// Issue an asynchronous request to flush all pending requests without
+  /// waiting for a full batch.
+  template <boost::asio::completion_token_for<void()> CompletionToken>
+  auto async_flush(CompletionToken&& token)
+  {
+    // maintain work on the default executor
+    auto consigned = boost::asio::consign(
+        std::forward<CompletionToken>(token),
+        boost::asio::make_work_guard(get_executor()));
+    return boost::asio::async_initiate<decltype(consigned), void()>(
+        initiate_async_flush{this}, consigned);
+  }
+
+ private:
+  MD5_HASH_CTX_MGR mgr;
+  executor_type ex;
+  boost::asio::steady_timer timer;
+  std::chrono::nanoseconds batch_timeout;
+
+  void complete(MD5_HASH_CTX* ctx, MD5_HASH_CTX* submit_ctx);
+
+  void flush(MD5_HASH_CTX* submit_ctx);
+
+  void init_async_hash(boost::asio::any_completion_handler<void(error_code)> handler,
+                       Digest& digest, std::string_view input, bool last);
+
+  struct initiate_async_hash {
+    Batch* self;
+
+    using executor_type = Batch::executor_type;
+    executor_type get_executor() const noexcept { return self->ex; }
```

Comment:
the intent of `struct initiate_async_hash` and `get_executor()` here was to run `init_async_hash()` within the Batch executor, but that doesn't seem to happen for 'awaitable' completion tokens

in the `isal_md5.batch_thread_pool` unit test, the calls to `init_async_hash()` are running on the thread pool context rather than the strand executor, hence the data races that lead to crashes